components/director/internal/dircleaner/service.go

```diff
@@ -0,0 +1,125 @@
+package dircleaner
+
+import (
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/kyma-incubator/compass/components/director/pkg/persistence"
+	"github.com/kyma-incubator/compass/components/director/pkg/tenant"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// CleanerService missing godoc
+type CleanerService interface {
+	Clean(ctx context.Context) error
+}
+
+// TenantService missing godoc
+type TenantService interface {
+	ListSubaccounts(ctx context.Context) ([]*model.BusinessTenantMapping, error)
+	GetTenantByID(ctx context.Context, id string) (*model.BusinessTenantMapping, error)
+	Update(ctx context.Context, id string, tenantInput model.BusinessTenantMappingInput) error
+	GetTenantByExternalID(ctx context.Context, id string) (*model.BusinessTenantMapping, error)
+	DeleteByExternalTenant(ctx context.Context, externalTenantID string) error
+}
+
+// CisService missing godoc
+type CisService interface {
+	GetGlobalAccount(ctx context.Context, subaccountID string) (string, error)
+}
+
+type service struct {
+	transact  persistence.Transactioner
+	tenantSvc TenantService
+	cisSvc    CisService
+}
+
+// NewCleaner missing godoc
+func NewCleaner(transact persistence.Transactioner, tenantSvc TenantService, cisSvc CisService) CleanerService {
+	return &service{
+		transact:  transact,
+		tenantSvc: tenantSvc,
+		cisSvc:    cisSvc,
+	}
+}
+
+// Clean missing godoc
+func (s *service) Clean(ctx context.Context) error {
+	tx, err := s.transact.Begin()
+	if err != nil {
+		return err
+	}
+	defer s.transact.RollbackUnlessCommitted(ctx, tx)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	log.C(ctx).Info("Listing all subaccounts provided by the event-service")
+	allSubaccounts, err := s.tenantSvc.ListSubaccounts(ctx)
+	if err != nil {
+		return errors.Wrap(err, "while getting all subaccounts")
+	}
+	log.C(ctx).Infof("Total number of listed subaccounts: %d", len(allSubaccounts))
+
+	for _, subaccount := range allSubaccounts {
+		log.C(ctx).Infof("Processing subaccount with ID %s", subaccount.ID)
+		globalAccountGUIDFromCis, err := s.cisSvc.GetGlobalAccount(ctx, subaccount.ExternalTenant)
+		if err != nil {
+			return errors.Wrapf(err, "while getting global account guid for subaacount with ID %s", subaccount.ID)
+		}
+		err = func() error {
+			tx, err := s.transact.Begin()
+			if err != nil {
+				return errors.Wrap(err, "failed to begin transaction")
+			}
+			ctx = persistence.SaveToContext(ctx, tx)
+			defer s.transact.RollbackUnlessCommitted(ctx, tx)
+
+			parentFromDB, err := s.tenantSvc.GetTenantByID(ctx, subaccount.Parent)
+			if err != nil {
+				log.C(ctx).Error(err)
+			}
+			if parentFromDB.ExternalTenant != globalAccountGUIDFromCis { // the record is directory and not GA
```

Comment:
if that's false (SA has valid parent) maybe we should log that the given SA is fine and doesn't need an update and increase the counter for successfully processed records